Resource metadata is remembered throughout the entire login flow.

halllo opened the pull request:
There currently is a problem with MCP servers that rely on a remote authorization server at a different host. During OAuth login the /authorize request is made to the correct authorization server, but the /token request is made against the MCP server.

Due to creating multiple transports and recursive invocations of `connectToRemoteServer()`, the acquired resource metadata (containing the correct authorization server) was forgotten between /authorize and /token calls.

Unfortunately the `transport` has the resource metadata in a private field, so I inject it by bypassing the typesystem. Going forward there should be only a single transport used for everything and no recursions. For now I consider this fix the bare minimum to get spec compliant OAuth to working at all.

halllo committed:
Due to creating multiple transports and recursive invocations, the acquired resource metadata (containing the correct authorization server) were forgotten between /authorize and /token calls. Now the metadata is remembered throughout the entire login flow.

clouatre reviewed (changes requested):
Reviewed and tested locally on macOS.

**Testing:**
- ✅ All tests pass (49/49)
- ✅ Build succeeds without errors
- ✅ Functionally solves the problem

**Request Changes:**

This PR uses a workaround when a simple proper fix exists.

**Current approach:**
- Creates two transports, discards the one with metadata
- Hacks private `_resourceMetadataUrl` field with `any` type
- Adds 283 lines of tests for the workaround
- Author admits: "Extremely hacky", "will break when refactored"

**Proper fix** (simpler, no hacks):

Instead of creating two transports and copying private fields, reuse the transport that discovered the metadata:

```typescript
// In utils.ts, around line 320-330:
if (!sseTransport && !client) {
  const testClient = new Client({ name: 'mcp-remote-test', version: '0.0.0' }, { capabilities: {} })
  await testClient.connect(transport) // transport discovers metadata during connection
  return transport // has metadata ✅
}
```

This eliminates:
- Private field access
- Type system bypass with `any`
- The need for 283-line test file
- The hack that "will break when refactored"

**Recommendation:**

Please implement the proper fix. It's simpler, cleaner, and doesn't require hacks or extensive test infrastructure.

halllo committed:
No more hacky injection and less recursion in success case. Support for OIDC metadata.

halllo commented:
## Resource metadata is now remembered throughout login

Fixed an issue where resource metadata (containing the correct authorization server) was lost between /authorize and /token calls due to multiple transports and recursive invocations.
Now, the discovered metadata is preserved and reused throughout the entire login flow, ensuring the correct authorization server is always used.
Reduced recursion and removed hacky injection logic in the authentication flow, making the process more robust and maintainable.
Added a comprehensive test (oauth.test.ts) to verify the login flow and token acquisition.

## Improved Authorization Server Metadata Discovery and OIDC support

The code now tries multiple well-known URLs, supporting both OAuth and OpenID Connect (OIDC) metadata discovery for broader compatibility (aligned with RFC 8414).

## Support for stateless MCP Servers

The client now probes MCP servers using both GET and POST methods when discovering the www-authenticate header. This ensures compatibility with stateless MCP servers that only support POST endpoints.

halllo commented:
Also addresses these issues:

- https://github.com/geelen/mcp-remote/issues/231
- https://github.com/geelen/mcp-remote/issues/214
- https://github.com/geelen/mcp-remote/issues/241
- https://github.com/geelen/mcp-remote/issues/207

mano1233 commented:
Is there any need for additional help with this pr? It would really help me setup my AgentCore MCP endpoint

halllo commented:
@mano1233 code-wise it is done and awaiting review and feedback.

I would highly appreciate, if you could clone this locally and confirm everything is working as expected. That should increase merge confidence.

You can run it locally like this:
`npx --package=/Users/Manuel.Naujoks/Projects/mcp-remote mcp-remote https://…`

mano1233 commented:
## Tested against AWS Bedrock AgentCore MCP endpoint with Okta OAuth

Ran a full sanity check of this PR against a production-like **AWS Bedrock AgentCore** MCP runtime that uses a **remote Okta custom authorization server** — the exact split-host scenario this PR fixes.

### Test Setup

| Component | Detail |
|-----------|--------|
| **MCP Server** | AWS Bedrock AgentCore runtime (stateless, POST-only) |
| **Auth Server** | Okta custom authorization server (separate host from MCP server) |
| **OAuth config** | `--static-oauth-client-info` with `openid email profile offline_access` scopes |
| **Transport** | `http-first` (default) → `StreamableHTTPClientTransport` |
| **Platform** | macOS, Node v20.19.3, built from this PR (v0.1.38 base) |

### Test Procedure

1. Cloned PR branch, ran `pnpm install && pnpm build` — clean, no errors
2. Ran full test suite — **110/110 tests pass** (5 test files)
3. Started proxy against AgentCore endpoint with `--debug`
4. Ran **end-to-end MCP protocol test**: spawned the proxy as a child process and exchanged JSON-RPC messages through STDIO

### Results

| Test | Status | Detail |
|------|--------|--------|
| Build & unit tests | ✅ | 110/110, clean build |
| GET/POST probing (#214) | ✅ | GET returned non-401; POST correctly returned 401 with `WWW-Authenticate` header. Critical for stateless MCP servers like AgentCore that only respond to POST |
| Protected Resource Metadata discovery | ✅ | Extracted `resource_metadata` URL from header, fetched PRM, resolved Okta auth server URL |
| OIDC metadata discovery | ✅ | `getMetadataUrls` generates both RFC 8414 and OIDC well-known URL candidates |
| Transport reuse (#231) | ✅ | Same transport instance reused — no metadata loss between test probe and actual connection |
| No recursion after auth | ✅ | Returns transport directly after `finishAuth` instead of recursive `connectToRemoteServer` |
| MCP `initialize` | ✅ | Server connected, protocol `2025-03-26`, capabilities: `[experimental, prompts, resources, tools]` |
| MCP `tools/list` | ✅ | Tools returned successfully |
| MCP `tools/call` | ✅ | Real query executed, 5115-char response with actual data returned |

### Debug Trace (abbreviated)

```
Probing MCP server for WWW-Authenticate header { method: 'GET' }
Probing MCP server for WWW-Authenticate header { method: 'POST' }
Received 401 with WWW-Authenticate header { hasHeader: true }
Successfully fetched Protected Resource Metadata { authorizationServers: ['<okta-auth-server>'] }
Using authorization server from Protected Resource Metadata
Connected to remote server using StreamableHTTPClientTransport
Proxy established successfully
```

### E2E Test Output

```
[1/3] INITIALIZE: PASS
[2/3] TOOLS/LIST: PASS
[3/3] TOOLS/CALL: PASS — response 5115 chars

============================================================
OVERALL: ALL TESTS PASSED
============================================================
```

### Summary

This PR **works end-to-end** for an AgentCore MCP + Okta setup. The GET/POST probing is essential for AgentCore (stateless, POST-only), and the transport-reuse approach cleanly solves the metadata-loss problem without requiring SDK API additions. Confirmed with a full MCP protocol exchange (`initialize` → `tools/list` → `tools/call`) returning real data through the proxy.
